Lean: Implement concurrency interface V2

ineol opened the pull request.

ineol committed:
Lean: Implement concurrency interface V2

tperami commented on src/sail_lean_backend/Sail/Sail.lean:
```diff
@@ -324,7 +350,13 @@ def Error.print : Error UE → String
   | User _ => "Uncaught user exception"
 
-section ConcurrencyInterface
+inductive Result (α : Type) (β : Type) where
+  | Ok (_ : α)
+  | Err (_ : β)
+  deriving Repr
```
This might be stupid question from a non-lean-user, but isn't there a standard library result type?

tperami commented on src/sail_lean_backend/Sail/Sail.lean:
```diff
@@ -383,7 +410,18 @@ structure Mem_write_request
+
+namespace ConcurrencyInterfaceV2
+
+structure Mem_request (n : Nat) (nt : Nat) (addr_size : Nat) (addr_space : Type) (mem_acc : Type) where
```
`n` and `nt` are type-level parameter in Sail but with constraints that `size` and `num_tag` have the same value due to Sail type-system constraints. In Rocq, `n` and `nt` do not exist and `size` and `num_tag` just contain the actual value. Would that be possible in Lean?

ineol replied:
In my copy of the rocq library I see 
```rocq
Record Mem_request {n : Z} {nt : Z} {addr_size : Z} {addr_space : Type} {mem_acc : Type}
(*(n >=? 0) && ((nt >=? 0) && (addr_size >? 0))*) := {
  Mem_request_access_kind : mem_acc;
  Mem_request_address : mword addr_size;
  Mem_request_address_space : addr_space;
  Mem_request_size : Z;
  Mem_request_num_tag : Z;
}.
```

tperami replied:
I'm looking at that one: https://github.com/rems-project/coq-sail/blob/61364217312039a86f5b71eea0d10ec58f5080a4/src-stdpp/ConcurrencyInterfaceV2.v#L126. Where is the other one coming from?

tperami replied:
@bacam I guess this is a question for you then, why is there two versions?

tperami replied:
So the reason for the 2 version is to have an as-if-sail-generated type and a nice-to-use type, and we're probably going to do the same in Lean, but the nice-to-use type is not needed in this PR

ineol committed:
Lean: compat between early return and exception

Early returns are encoded using an exception monad in Lean, and it did
not allow to throw exceptions. As a by-product, we move some definitions
from `Specialization.lean` to `Sail.lean`.

tperami commented:
LGTM

bacam commented:
I expected the `concurrency_interface_v2` test to work with this, but I'm getting some errors where Lean is unable to find typeclass instances.  Have you looked at this test?

ineol commented:
@bacam thank you for the feedback! No I missed it. I fixed compilation but the behavior of the test isn't working yet. Partially because we currently ignore the tags in the state monad, but maybe also for other reasons...

bacam commented:
It's only the tag part of the test that's failing, the plain memory accesses and other operations appear to be fine.  Shall I merge it, or would you like to add tag accesses now?

bacam commented:
I just gave this a spin on a version of the RISC-V model that uses v2 (https://github.com/bacam/sail-riscv/tree/archsem2), and the only problem I had was that the `access_kind` type wasn't opened, causing errors in the functions used to instantiate `Arch`.  (I also had to tweak `RiscvExtras.lean`, but I was expecting that.)

ineol commented:
I fixed the `access_kind` bug (I think). I think it makes sense to merge it if your build of the risc-v model goes through.

bacam commented:
Yes, it builds now.
